l2geth/core/evm.go

```diff
@@ -66,12 +66,13 @@ func NewEVMContext(msg Message, header *types.Header, chain ChainContext, author
 			GetHash:         GetHashFn(header, chain),
 			Origin:          msg.From(),
 			Coinbase:        beneficiary,
-			BlockNumber:     msg.L1BlockNumber(),
+			BlockNumber:     new(big.Int).Set(header.Number),
 			Time:            new(big.Int).SetUint64(msg.L1Timestamp()),
 			Difficulty:      new(big.Int).Set(header.Difficulty),
 			GasLimit:        header.GasLimit,
 			GasPrice:        new(big.Int).Set(msg.GasPrice()),
 			L1MessageSender: l1MessageSender,
+			L1BlockNumber:   msg.L1BlockNumber(),
```

Comment:
Doing it this way couples the L1 block number to the txmeta. This is ok for now but we should try to figure out how to eliminate the need for the txmeta. I think the proposed "context tx type" should be able to do this.